### PR TITLE
Implement daemon.Write

### PR DESCRIPTION
**Pull request by nkubala**
This adds support for loading a `v1.Image` into a local docker daemon, by first writing it in `docker save` format using `tarball.Write`, and then feeding that to the docker client.

Previously `tarball.Write` only supported writing to disk, so to avoid this I changed the signature to return an `io.ReadCloser` by default. `tarball.WriteToFile` can be used in its place to write the contents of this reader to disk.

### CODE REVIEW

File: v1/tarball/write.go [hunk 1]
```diff
@@ -49,6 +59,37 @@ func Write(p string, tag name.Tag, img v1.Image, wo *WriteOptions) error {
...
+// Return a reader for contents of a tarball, containing:
+// One manifest.json file at the top level containing information about several images.
+// One file for each layer, named after the layer's SHA.
+// One file for the config blob, named after its SHA.
+func write(tag name.Tag, img v1.Image, wo *WriteOptions, w io.Writer) error {
+	tf := tar.NewWriter(w)
+	defer tf.Close()
+
+	if img == nil {
+		return fmt.Errorf("wtf we have a serious problem")
```
[the remark this answers is not in the document]
Comment (nkubala) [reply v1/tarball/write.go]: 😬

File: v1/tarball/write.go [hunk 2]
```diff
@@ -32,14 +32,24 @@ type WriteOptions struct {
...
-	// One file for each layer, named after the layer's SHA.
-	// One file for the config blob, named after its SHA.
+// Write writes the image at the given tag to a tarball,
+// and returns it through a reader to be consumed
+func Write(tag name.Tag, img v1.Image, wo *WriteOptions) io.ReadCloser {
+	pr, pw := io.Pipe()
+
+	go func() {
+		// Close the writer with any errors encountered during
+		// extraction. These errors will be returned by the reader end
```
Comment (ImJasonH) [inline v1/tarball/write.go]: This comment mentions extraction.

File: v1/tarball/write.go [hunk 3]
```diff
@@ -49,6 +59,33 @@ func Write(p string, tag name.Tag, img v1.Image, wo *WriteOptions) error {
 	tf := tar.NewWriter(w)
 	defer tf.Close()
 
+	tr := tar.NewReader(Write(tag, img, wo))
```
Comment (ImJasonH) [inline v1/tarball/write.go]: You should be able to just `io.Copy(f, r)` I think right?

File: v1/daemon/write.go [hunk 4]
```diff
@@ -15,19 +15,45 @@
...
 // WriteOptions are used to expose optional information to guide or
 // control the image write.
 type WriteOptions struct {
 	// TODO(dlorenc): What kinds of knobs does the daemon expose?
 }
 
 // Write saves the image into the daemon as the given reference.
-func Write(ref name.Reference, img v1.Image, wo WriteOptions) error {
-	return fmt.Errorf("NYI: daemon.Write(%v)", ref)
+func Write(ref name.Reference, img v1.Image, wo WriteOptions) (types.ImageLoadResponse, error) {
```
Comment (ImJasonH) [inline v1/daemon/write.go]: This library's API shouldn't return some other library's types. Is there another way to express this?

File: v1/tarball/write.go (hunk 2 above)
Comment (nkubala) [reply v1/tarball/write.go]: thanks

File: v1/tarball/write.go (hunk 3 above)
Comment (nkubala) [reply v1/tarball/write.go]: I thought so at first, but the copy doesn't seem to read correctly. Not sure if this is because it's a tar reader or if it's coming from a pipe, but if you have suggestions I'm all ears

File: v1/daemon/write.go (hunk 4 above)
Comment (nkubala) [reply v1/daemon/write.go]: ah yeah that's a great point. I still want to expose the client's response to the user, so I'll change this to return `(string, error)`

File: v1/daemon/write.go [hunk 5]
```diff
@@ -15,19 +15,53 @@
...
+		return "", err
+	}
+	// write the image in docker save format first, then load it
+	reader := tarball.Write(tag, img, &tarball.WriteOptions{})
+	resp, err := cli.ImageLoad(context.Background(), reader, false)
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	resp.Body.Close()
```
Comment (dlorenc) [inline v1/daemon/write.go]: nit: move this up as a defer before you do the ReadAll. Also check the err from ImageLoad separately.

File: v1/daemon/write_test.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/google/go-containerregistry/name"
+	"github.com/google/go-containerregistry/v1/tarball"
+)
+
+func TestWriteImage(t *testing.T) {
```
Comment (jonjohnsonjr) [inline v1/daemon/write_test.go]: This doesn't override `getImageLoader`, so it has a dependency on docker, which I don't have installed, so I can't run the tests :)

```
write_test.go:36: Error writing image tar: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

(you probably have some `test_image_2:latest`s in your daemon 😂 )

File: v1/tarball/write.go [hunk 7]
```diff
@@ -32,14 +32,24 @@ type WriteOptions struct {
...
-// Write saves the image as the given tag in a tarball at the given path.
-func Write(p string, tag name.Tag, img v1.Image, wo *WriteOptions) error {
-	// Write in the compressed format.
-	// This is a tarball, on-disk, with:
-	// One manifest.json file at the top level containing information about several images.
-	// One file for each layer, named after the layer's SHA.
-	// One file for the config blob, named after its SHA.
+// Write writes the image at the given tag to a tarball,
+// and returns it through a reader to be consumed
+func Write(tag name.Tag, img v1.Image, wo *WriteOptions) io.ReadCloser {
```
Comment (jonjohnsonjr) [inline v1/tarball/write.go]: It's weird to me that something called `Write` returns a `ReadCloser`. After spending a few minutes trying to come up with a name that made more sense, I think I realized the issue.

This would be a lot cleaner if `tarball.Write` was instead:
```go
func Write(w io.Writer, tag name.Tag, img v1.Image, wo *WriteOptions) error
```

(Hey look, that's `tarball.write` ;P)

Then `tarball.WriteToFile` could just be sugar to create the file for you:
```go
func WriteToFile(p string, tag name.Tag, img v1.Image, wo *WriteOptions) error {
	f, err := os.Create(p)
	if err != nil {
		return err
	}
	defer f.Close()

	return Write(tag, img, wo, f)
}
```

And then you can move the `io.Pipe` from `tarball.Write` into `daemon.Write`, which would just pipe `tarball.Write` into `cli.ImageLoad`.

@ImJasonH thoughts? I think this would be a bit more idiomatic/composable.

File: v1/daemon/write.go [hunk 8]
```diff
@@ -15,19 +15,53 @@
...
+	}
+
+	tag, err := name.NewTag(ref.Name(), name.WeakValidation)
+	if err != nil {
+		return "", err
+	}
+	// write the image in docker save format first, then load it
+	reader := tarball.Write(tag, img, &tarball.WriteOptions{})
+	resp, err := cli.ImageLoad(context.Background(), reader, false)
+	b, err := ioutil.ReadAll(resp.Body)
```
Comment (jonjohnsonjr) [inline v1/daemon/write.go]: It might be better to just make `daemon.Write` return an `io.ReadCloser` (i.e. `resp.Body`) instead of a string? I'm not sure how this will be used, so it's hard to say.

File: v1/tarball/write.go (hunk 7 above)
Comment (ImJasonH) [reply v1/tarball/write.go]: Yes! I was bumping up on that too and just didn't think about it enough. I think we could probably port that same change to `mutate.Extract` too (why is `Extract` in `mutate`?)

But yeah that's definitely a better API

Comment (jonjohnsonjr) [reply v1/tarball/write.go]: +1 to giving `mutate.Extract` the same treatment.

Disclaimer: the rest of this comment is unrelated to this PR.

tl;dr I think we need `fs.Write` and `fs.Image` (or `fs.Layer`)

> (why is Extract in mutate?)

I'll bring up [my comment](https://github.com/google/go-containerregistry/pull/99#issuecomment-383757347) again -- if we can change `mutate.Extract` to:

```go
func Flatten(v1.Image) (v1.Image, error)
```

... I think it makes sense to leave it in `mutate`. Otherwise, I agree it doesn't really fit the style of mutations in there (maybe it belongs in `v1util` or `tarball`? or a new `fs.Write`?). Unfortunately, the change I want makes things a bit complicated and potentially a lot slower...

I'm wondering if we can abuse this to get what I want without much code... e.g.:

```go
func Flatten(i v1.Image) (v1.Image, error) {
  // pseudo-go, will likely need lots of plumbing to avoid excessive buffering
  return mutate.AppendLayers(empty.Image(), layer.Uncompressed(mutate.Extract(i)))
}
```

Then implementing `{remote,tarball,daemon} | flatten | {remote,tarball,daemon}` is trivial :)

File: v1/tarball/write.go (hunk 3 above)
Comment (nkubala) [reply v1/tarball/write.go]: -1 for the pun 😆 

you're right about the extra read/write wrapping, but yeah this definitely wasn't the best construction. fixed up to make a little bit more sense

File: v1/daemon/write.go [hunk 9]
```diff
@@ -15,19 +15,53 @@
...
 // Write saves the image into the daemon as the given reference.
-func Write(ref name.Reference, img v1.Image, wo WriteOptions) error {
-	return fmt.Errorf("NYI: daemon.Write(%v)", ref)
+func Write(ref name.Reference, img v1.Image, wo WriteOptions) (string, error) {
+	cli, err := getImageLoader()
+	if err != nil {
+		return "", err
+	}
+
+	tag, err := name.NewTag(ref.Name(), name.WeakValidation)
```
[the remark this answers is not in the document]
Comment (nkubala) [reply v1/daemon/write.go]: not sure why I did this....maybe I was just modeling after `remote.Write`? now that you mention it this probably should handle digests, but then shouldn't `tarball.Write` as well?

File: v1/tarball/write.go [hunk 10]
```diff
@@ -32,14 +32,24 @@ type WriteOptions struct {
...
 }
 
-// Write saves the image as the given tag in a tarball at the given path.
-func Write(p string, tag name.Tag, img v1.Image, wo *WriteOptions) error {
-	// Write in the compressed format.
-	// This is a tarball, on-disk, with:
-	// One manifest.json file at the top level containing information about several images.
-	// One file for each layer, named after the layer's SHA.
-	// One file for the config blob, named after its SHA.
+// Write writes the image at the given tag to a tarball,
```
[the remark this answers is not in the document]
Comment (nkubala) [reply v1/tarball/write.go]: yeah fair enough. I would have clarified but I removed this altogether :)

File: v1/daemon/write_test.go (hunk 6 above)
Comment (nkubala) [reply v1/daemon/write_test.go]: this is testing loading an image into the local daemon, so if you don't have docker installed this will definitely fail :)

the test is just doing a `tarball.ImageFromPath` on one of the tars checked into testdata and then retagging it as `test_image_2:latest`.

File: v1/tarball/write.go (hunk 7 above)
Comment (nkubala) [reply v1/tarball/write.go]: Thanks for the feedback here, this is definitely a better construction. Passing in a writer rather than returning a reader puts the onus on the caller to do the IO, which I think is a lot cleaner.

re `Extract` being in `mutate`: I originally had this in `tarball`, but wasn't sure it made sense in there since it's not technically limited to `tarball`s (though they do get converted before doing the actual extraction).

The actual motivation for this was twofold: I wanted to have a cleaner way of doing filesystem comparisons in container-diff (extracting the flattened fs is an obvious choice), but then I also wanted to implement `Flatten` in its original state because it's cool :) 

I tried to make it as extensible as possible, and I guess `mutate` was the only neutral package that made sense. I'm fine moving it around though.

**Comment by nkubala** [comment]
this should be RFAL

File: v1/daemon/write_test.go (hunk 6 above)
Comment (jonjohnsonjr) [reply v1/daemon/write_test.go]: I won't fight this too hard, but it would be nice if I could `bazel test ...` this without installing docker :) can you mock out the docker interaction for this test and add an integration test that exercises the actual daemon?

I'm not a huge fan of implicit dependencies, but if nobody else cares, it's nbd.

File: v1/daemon/write.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonjohnsonjr) [reply v1/daemon/write.go]: I think something about the tarball format expects a tag (or used to?). Not sure.

File: v1/daemon/write.go [hunk 11]
```diff
@@ -15,19 +15,58 @@
...
+	go func() {
+		pw.CloseWithError(tarball.Write(tag, img, &tarball.WriteOptions{}, pw))
+	}()
+
+	// write the image in docker save format first, then load it
+	resp, err := cli.ImageLoad(context.Background(), pr, false)
+	defer resp.Body.Close()
+	b, readErr := ioutil.ReadAll(resp.Body)
+	response := string(b)
+	if err != nil {
```
Comment (jonjohnsonjr) [inline v1/daemon/write.go]: I think you need to check this before `defer resp.Body.Close()` , since resp might be nil.

Comment (nkubala) [reply v1/daemon/write.go]: ah yeah good catch, resp can't be nil but `resp.Body` can

File: v1/daemon/write_test.go (hunk 6 above)
Comment (nkubala) [reply v1/daemon/write_test.go]: no that's a good point, if we were relying on docker to always be present then what would be the point of this library :) fixed to mock out so we don't need docker to test

File: v1/tarball/write.go [hunk 12]
```diff
@@ -32,20 +32,24 @@ type WriteOptions struct {
...
 	if err != nil {
 		return err
 	}
 	defer w.Close()
 
+	return Write(tag, img, wo, w)
+}
+
+// Write in the compressed format.
+// Return a reader for contents of a tarball, containing:
```
Comment (jonjohnsonjr) [inline v1/tarball/write.go]: Update this comment.

File: v1/daemon/write.go [hunk 13]
```diff
@@ -15,19 +15,58 @@
...
+	return client.NewEnvClient()
+}
+
 // WriteOptions are used to expose optional information to guide or
 // control the image write.
 type WriteOptions struct {
 	// TODO(dlorenc): What kinds of knobs does the daemon expose?
 }
 
 // Write saves the image into the daemon as the given reference.
```
Comment (jonjohnsonjr) [inline v1/daemon/write.go]: s/reference/tag/

File: v1/daemon/write_test.go (hunk 6 above)
Comment (jonjohnsonjr) [reply v1/daemon/write_test.go]: Thanks!
